src/examples/algorithms/AnomalyDetection.cpp

```diff
@@ -0,0 +1,221 @@
+
+class AnomalyDetection
+{
+
+  private:
+    ScalarEncoder encoder;
+    SpatialPooler sp;
+    Cells4 tp;
+    TemporalMemory tm;
+    Anomaly anomaly; 
+    std::vector<UInt> lastTPOutput_;
+
+  public:
+    Real compute(Real x) {
+        auto scalar = std::vector<Real32>(encoder.getOutputWidth());
+        encoder.encodeIntoArray(x, scalar.data());
+
+        // For some reason, the ScalarEncoder outputs a Real32 vector,
+        // which is incomaptible with the SpatialPooler's input, hence a new
+        // vector is created with the required type, and the elements
+        // are converted.
+        auto uint_scalar = VectorHelpers::castVectorType<Real32, UInt>(scalar);
+        if (DEBUG_LEVEL > 1) {
+          std::cout << "Scalar encoder: ";
+          VectorHelpers::print_vector(uint_scalar);
+        }
+
+        std::vector<UInt> spOutput(sp.getNumColumns());
+        sp.compute(uint_scalar.data(), true, spOutput.data());
+        sp.stripUnlearnedColumns(spOutput.data());
+        if (DEBUG_LEVEL > 2) {
+          std::cout << "Spatial pooler: ";
+          VectorHelpers::print_vector(VectorHelpers::binaryToSparse<UInt>(spOutput), ",");
+          std::cout << std::endl;
+        }
+
+        // The temporal pooler uses Real32, so the vector must be converted again
+        auto tpInput = VectorHelpers::castVectorType<UInt, Real>(spOutput);
+        auto nCells = 0;
+        if (tmImpl == "TP") {
+          nCells = tp.nCells(); //TODO fix API of (SP)/TM/TP to same format, so this can be avoided
+        else { 
+          nCells = tm.numberOfCells();
+        }
+        std::vector<Real> tpOutput(nCells);
+        if (tmImpl == "TP") {
+          tp.compute(tpInput.data(), tpOutput.data(), true, true);
+        else {
+          tm.compute(tpInput.size(), tpInput.data(), true);
+          tpOutput = tm.getActiveCells(); //FIXME do union with getPredictedCells() , like TP.outputMode="both"
+        }
+
+        // And the result is converted ONCE again to UInts for pretty printing to stdout.
+        auto uintTpOutput = VectorHelpers::castVectorType<Real32, UInt>(tpOutput);
+        if (DEBUG_LEVEL > 3) {
+          std::cout << "Output of temporal pooler: ";
+          VectorHelpers::print_vector(VectorHelpers::binaryToSparse<UInt>(uintTpOutput), ",");
+          std::cout << std::endl;
+        }
+
+        Real anScore = anomaly.compute(
+          VectorHelpers::binaryToSparse<UInt>(spOutput),     // Spatial pooler (current active columns)
+          VectorHelpers::binaryToSparse<UInt>(lastTPOutput_), // Temporal pooler (previously predicted columns)
+          0,
+          0
+        );
+        // Save the output of the TP for the next iteration...
+        if (tmImpl == "TP") {
+          lastTPOutput_ = VectorHelpers::cellsToColumns(uintTpOutput, tp.nCellsPerCol());
+        else {
+          lastTPOutput_ = VectorHelpers::cellsToColumns(uintTpOutput, tm.getCellsPerCol());
```

Comment:
The TM has a `columnForCell` method. That should be sufficient. Anyone who wants a `cellsToColumns` helper function can implement it using `columnForCell`. This helper function might return a vector, or a set, or a sorted_set, or some other custom data structure.

I intentionally chose not to have a `cellsToColumns` method. I exposed the minimal amount needed so that people could implement whatever version of `cellsToColumns` function they want.